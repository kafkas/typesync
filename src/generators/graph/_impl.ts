import { schema } from '../../schema/index.js';
import { assertNever } from '../../util/assert.js';
import type { GraphGeneration, GraphGenerator, GraphGeneratorConfig } from './_types.js';
import { MermaidGraph, MermaidGraphNode, MermaidGraphOrientation } from './mermaid-graph.js';
import { Collection, Document, SchemaGraph } from './schema-graph/interfaces.js';

type SchemaGraphOrientation = 'vertical' | 'horizontal';

// function isGeneric(part: string) {
//   return part.startsWith('{') && part.endsWith('}');
// }

export class GraphGeneratorImpl implements GraphGenerator {
  public constructor(private readonly config: GraphGeneratorConfig) {}

  public generate(s: schema.Schema): GraphGeneration {
    const { documentModels: _ } = s;
    const schemaGraph = this.buildSchemaGraphFromSchema(s);
    const mermaidGraph = this.buildMermaidGraphFromSchemaGraph(schemaGraph);
    return {
      type: 'graph',
      graph: mermaidGraph,
    };
  }

  public buildSchemaGraphFromSchema(_s: schema.Schema): SchemaGraph {
    // TODO: Implement
    return { children: { type: 'literal-graph-children', collections: [] } };
  }

  public buildMermaidGraphFromSchemaGraph(graph: SchemaGraph): MermaidGraph {
    const mermaidGraphOrientation = this.getMermaidGraphOrientation(this.config.orientation);
    const mermaidGraph = new MermaidGraph(mermaidGraphOrientation);
    if (graph.children.type === 'generic-graph-children') {
      this.buildNodeForCollection(mermaidGraph, graph.children.collection);
    } else if (graph.children.type === 'literal-graph-children') {
      graph.children.collections.forEach(col => {
        this.buildNodeForCollection(mermaidGraph, col);
      });
    } else {
      assertNever(graph.children);
    }
    return mermaidGraph;
  }

  private buildNodeForCollection(mermaidGraph: MermaidGraph, col: Collection): MermaidGraphNode {
    const colNode = mermaidGraph.createNode(col.id);
    if (col.children.type === 'generic-collection-children') {
      const docNode = this.buildNodeForDocument(mermaidGraph, col.children.document);
      mermaidGraph.link(colNode, docNode);
    } else if (col.children.type === 'literal-collection-children') {
      col.children.documents.forEach(doc => {
        const docNode = this.buildNodeForDocument(mermaidGraph, doc);
        mermaidGraph.link(colNode, docNode);
      });
    } else {
      assertNever(col.children);
    }
    return colNode;
  }

  private buildNodeForDocument(mermaidGraph: MermaidGraph, doc: Document): MermaidGraphNode {
    const docNode = mermaidGraph.createNode(doc.id);
    if (doc.children !== null) {
      if (doc.children.type === 'generic-document-children') {
        const colNode = this.buildNodeForCollection(mermaidGraph, doc.children.collection);
        mermaidGraph.link(docNode, colNode);
      } else if (doc.children.type === 'literal-document-children') {
        doc.children.collections.forEach(col => {
          const colNode = this.buildNodeForCollection(mermaidGraph, col);
          mermaidGraph.link(docNode, colNode);
        });
      } else {
        assertNever(doc.children);
      }
    }
    return docNode;
  }

  private getMermaidGraphOrientation(orientation: SchemaGraphOrientation): MermaidGraphOrientation {
    switch (orientation) {
      case 'vertical':
        return 'TB';
      case 'horizontal':
        return 'LR';
      default:
        assertNever(orientation);
    }
  }
}

export function createGraphGenerator(config: GraphGeneratorConfig): GraphGenerator {
  return new GraphGeneratorImpl(config);
}
