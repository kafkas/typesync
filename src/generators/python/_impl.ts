import { python } from '../../platforms/python/index.js';
import { Schema } from '../../schema/index.js';
import { assertNever } from '../../util/assert.js';
import { adjustSchemaForPython } from './_adjust-schema.js';
import { flatTypeToPython } from './_converters.js';
import type {
  PythonAliasDeclaration,
  PythonDeclaration,
  PythonEnumClassDeclaration,
  PythonGeneration,
  PythonGenerator,
  PythonGeneratorConfig,
  PythonPydanticClassDeclaration,
} from './_types.js';

class PythonGeneratorImpl implements PythonGenerator {
  public constructor(private readonly config: PythonGeneratorConfig) {}

  public generate(s: Schema): PythonGeneration {
    const adjustedSchema = adjustSchemaForPython(s);
    const { aliasModels, documentModels } = adjustedSchema;
    const declarations: PythonDeclaration[] = [];
    aliasModels.forEach(model => {
      const d = this.createDeclarationForFlatAliasModel(model);
      declarations.push(d);
    });
    documentModels.forEach(model => {
      const d = this.createDeclarationForFlatDocumentModel(model);
      declarations.push(d);
    });
    return { type: 'python', declarations };
  }

  private createDeclarationForFlatAliasModel(model: python.schema.AliasModel): PythonDeclaration {
    switch (model.type.type) {
      case 'unknown':
      case 'nil':
      case 'string':
      case 'boolean':
      case 'int':
      case 'double':
      case 'timestamp':
      case 'string-literal':
      case 'int-literal':
      case 'boolean-literal':
      case 'tuple':
      case 'list':
      case 'map':
      case 'discriminated-union':
      case 'simple-union':
      case 'alias':
        return this.createDeclarationForFlatType(model.type, model.name, model.docs);
      case 'string-enum':
      case 'int-enum':
        return this.createDeclarationForEnumType(model.type, model.name, model.docs);
      case 'object':
        return this.createDeclarationForFlatObjectType(model.type, model.name, model.docs);
      default:
        assertNever(model.type);
    }
  }

  private createDeclarationForFlatDocumentModel(model: python.schema.DocumentModel): PythonDeclaration {
    // A Firestore document can be considered an 'object' type
    return this.createDeclarationForFlatObjectType(model.type, model.name, model.docs);
  }

  private createDeclarationForEnumType(
    type: python.schema.types.Enum,
    modelName: string,
    modelDocs: string | null
  ): PythonEnumClassDeclaration {
    const pythonType: python.EnumClass = {
      type: 'enum-class',
      attributes: type.members.map(item => ({
        key: item.label,
        value: item.value,
      })),
    };
    return {
      type: 'enum-class',
      modelName,
      modelType: pythonType,
      modelDocs,
    };
  }

  private createDeclarationForFlatObjectType(
    type: python.schema.types.Object,
    modelName: string,
    modelDocs: string | null
  ): PythonPydanticClassDeclaration {
    const pythonType: python.ObjectClass = {
      type: 'object-class',
      attributes: type.fields.map(f => ({
        name: f.name,
        type: flatTypeToPython(f.type),
        docs: f.docs,
        optional: f.optional,
      })),
      additionalAttributes: type.additionalFields,
    };
    return {
      type: 'pydantic-class',
      modelName,
      modelType: pythonType,
      modelDocs,
    };
  }

  private createDeclarationForFlatType(
    type: python.schema.types.Type,
    modelName: string,
    modelDocs: string | null
  ): PythonAliasDeclaration {
    const pythonType = flatTypeToPython(type);
    return {
      type: 'alias',
      modelName,
      modelType: pythonType,
      modelDocs,
    };
  }
}

export function createPythonGenerator(config: PythonGeneratorConfig): PythonGenerator {
  return new PythonGeneratorImpl(config);
}
