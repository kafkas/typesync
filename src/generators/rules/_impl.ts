import { schema } from '../../schema/index.js';
import { adjustSchemaForRules } from './_adjust-schema.js';
import { flatObjectTypeToRules, flatTypeToRules } from './_converters.js';
import type {
  RulesGeneration,
  RulesGenerator,
  RulesGeneratorConfig,
  RulesReadonlyFieldValidatorDeclaration,
  RulesTypeValidatorDeclaration,
} from './_types.js';

class RulesGeneratorImpl implements RulesGenerator {
  public constructor(private readonly config: RulesGeneratorConfig) {}

  public generate(s: schema.Schema): RulesGeneration {
    const adjustedSchema = adjustSchemaForRules(s);
    const { aliasModels, documentModels } = adjustedSchema;
    const typeValidatorDeclarations: RulesTypeValidatorDeclaration[] = [
      ...aliasModels.map(model => this.createTypeValidatorDeclarationForFlatAliasModel(model.name, model.type)),
      ...documentModels.map(model => this.createTypeValidatorDeclarationForFlatDocumentModel(model.name, model.type)),
    ];
    const readonlyFieldValidatorDeclarations: RulesReadonlyFieldValidatorDeclaration[] = [
      ...aliasModels
        .map(model => ({ modelName: model.name, modelType: model.type }))
        .filter(
          (params): params is { modelName: string; modelType: schema.rules.types.Object } =>
            params.modelType.type === 'object'
        )
        .map(params =>
          this.createReadonlyFieldValidatorDeclarationForFlatAliasModel(params.modelName, params.modelType)
        ),
      ...documentModels.map(model =>
        this.createReadonlyFieldValidatorDeclarationForFlatDocumentModel(model.name, model.type)
      ),
    ];
    return { type: 'rules', typeValidatorDeclarations, readonlyFieldValidatorDeclarations };
  }

  private createTypeValidatorDeclarationForFlatAliasModel(
    modelName: string,
    modelType: schema.rules.types.Type
  ): RulesTypeValidatorDeclaration {
    const rulesType = flatTypeToRules(modelType);
    return {
      type: 'type-validator',
      modelName,
      modelType: rulesType,
    };
  }

  private createTypeValidatorDeclarationForFlatDocumentModel(
    modelName: string,
    modelType: schema.rules.types.Object
  ): RulesTypeValidatorDeclaration {
    const rulesType = flatObjectTypeToRules(modelType);
    return {
      type: 'type-validator',
      modelName,
      modelType: rulesType,
    };
  }

  private createReadonlyFieldValidatorDeclarationForFlatAliasModel(
    modelName: string,
    _modelType: schema.rules.types.Object
  ): RulesReadonlyFieldValidatorDeclaration {
    return {
      type: 'readonly-field-validator',
      modelName,
    };
  }

  private createReadonlyFieldValidatorDeclarationForFlatDocumentModel(
    modelName: string,
    _modelType: schema.rules.types.Object
  ): RulesReadonlyFieldValidatorDeclaration {
    return {
      type: 'readonly-field-validator',
      modelName,
    };
  }
}

export function createRulesGenerator(config: RulesGeneratorConfig): RulesGenerator {
  return new RulesGeneratorImpl(config);
}
