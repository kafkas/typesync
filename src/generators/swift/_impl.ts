import { MixedEnumValueTypesNotSupportedError } from '../../errors/generator.js';
import { swift } from '../../platforms/swift/index.js';
import { schema } from '../../schema/index.js';
import { assertNever } from '../../util/assert.js';
import { flatTypeToSwift } from './_converters.js';
import { flattenSchema } from './_flatten-schema.js';
import {
  FlatAliasModel,
  FlatDiscriminatedUnionType,
  FlatDocumentModel,
  FlatObjectType,
  FlatSimpleUnionType,
  FlatType,
} from './_schema.js';
import type {
  SwiftDeclaration,
  SwiftEnumWithAssociatedValuesDeclaration,
  SwiftGeneration,
  SwiftGenerator,
  SwiftGeneratorConfig,
  SwiftIntEnumDeclaration,
  SwiftStringEnumDeclaration,
  SwiftStructDeclaration,
  SwiftTypealiasDeclaration,
} from './_types.js';

class SwiftGeneratorImpl implements SwiftGenerator {
  public constructor(private readonly config: SwiftGeneratorConfig) {}

  public generate(s: schema.Schema): SwiftGeneration {
    const flattenedSchema = flattenSchema(s);
    const { aliasModels, documentModels } = flattenedSchema;
    const declarations: SwiftDeclaration[] = [];
    aliasModels.forEach(model => {
      const d = this.createDeclarationForFlatAliasModel(model);
      declarations.push(d);
    });
    documentModels.forEach(model => {
      const d = this.createDeclarationForFlatDocumentModel(model);
      declarations.push(d);
    });
    return { type: 'swift', declarations };
  }

  private createDeclarationForFlatAliasModel(model: FlatAliasModel): SwiftDeclaration {
    switch (model.type.type) {
      case 'nil':
      case 'string':
      case 'boolean':
      case 'int':
      case 'double':
      case 'timestamp':
      case 'literal':
      case 'tuple':
      case 'list':
      case 'map':
      case 'alias':
        return this.createDeclarationForFlatType(model.type, model.name, model.docs);
      case 'enum':
        return this.createDeclarationForEnumType(model.type, model.name, model.docs);
      case 'object':
        return this.createDeclarationForFlatObjectType(model.type, model.name, model.docs);
      case 'discriminated-union':
        return this.createDeclarationForFlatDiscriminatedUnionType(model.type, model.name, model.docs);
      case 'simple-union':
        return this.createDeclarationForFlatSimpleUnionType(model.type, model.name, model.docs);
      default:
        assertNever(model.type);
    }
  }

  private createDeclarationForFlatDocumentModel(model: FlatDocumentModel): SwiftDeclaration {
    // A Firestore document can be considered an 'object' type
    return this.createDeclarationForFlatObjectType(model.type, model.name, model.docs);
  }

  private createDeclarationForEnumType(
    type: schema.types.Enum,
    modelName: string,
    modelDocs: string | undefined
  ): SwiftStringEnumDeclaration | SwiftIntEnumDeclaration {
    const isStringEnum = type.members.every(member => typeof member.value === 'string');
    if (isStringEnum) {
      const swiftType: swift.StringEnum = {
        type: 'string-enum',
        cases: type.members.map(member => ({
          key: member.label,
          value: member.value as string,
        })),
      };
      return {
        type: 'string-enum',
        modelType: swiftType,
        modelName,
        modelDocs,
      };
    }

    const isIntEnum = type.members.every(member => typeof member.value === 'number');
    if (isIntEnum) {
      const swiftType: swift.IntEnum = {
        type: 'int-enum',
        cases: type.members.map(member => ({
          key: member.label,
          value: member.value as number,
        })),
      };
      return {
        type: 'int-enum',
        modelType: swiftType,
        modelName,
        modelDocs,
      };
    }

    throw new MixedEnumValueTypesNotSupportedError(modelName);
  }

  private createDeclarationForFlatObjectType(
    type: FlatObjectType,
    modelName: string,
    modelDocs: string | undefined
  ): SwiftStructDeclaration {
    const swiftType: swift.Struct = {
      type: 'struct',
      properties: type.fields.map(f => ({
        name: f.name,
        type: flatTypeToSwift(f.type),
        docs: f.docs,
        optional: f.optional,
      })),
    };
    return {
      type: 'struct',
      modelName,
      modelType: swiftType,
      modelDocs,
    };
  }

  private createDeclarationForFlatDiscriminatedUnionType(
    _type: FlatDiscriminatedUnionType,
    _modelName: string,
    _modelDocs: string | undefined
  ): SwiftEnumWithAssociatedValuesDeclaration {
    // TODO: Implement
    throw new Error('Unimplemented');
  }

  private createDeclarationForFlatSimpleUnionType(
    _type: FlatSimpleUnionType,
    _modelName: string,
    _modelDocs: string | undefined
  ): SwiftEnumWithAssociatedValuesDeclaration {
    // TODO: Implement
    throw new Error('Unimplemented');
  }

  private createDeclarationForFlatType(
    type: FlatType,
    modelName: string,
    modelDocs: string | undefined
  ): SwiftTypealiasDeclaration {
    const swiftType = flatTypeToSwift(type);
    return {
      type: 'typealias',
      modelName,
      modelType: swiftType,
      modelDocs,
    };
  }
}

export function createSwiftGenerator(config: SwiftGeneratorConfig): SwiftGenerator {
  return new SwiftGeneratorImpl(config);
}
