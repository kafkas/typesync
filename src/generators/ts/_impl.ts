import { converters } from '../../converters';
import { schema } from '../../schema';
import type { TSDeclaration, TSGeneration, TSGenerator, TSGeneratorConfig } from './_types';

class TSGeneratorImpl implements TSGenerator {
  public constructor(private readonly config: TSGeneratorConfig) {}

  public generate(s: schema.Schema): TSGeneration {
    const { aliasModels, documentModels } = s;
    const declarations: TSDeclaration[] = [];

    aliasModels.forEach(model => {
      const tsType = converters.schema.typeToTS(model.value);
      declarations.push({ type: 'alias', modelName: model.name, modelType: tsType });
    });

    documentModels.forEach(model => {
      // A Firestore document can be considered an 'object' type
      const tsType = converters.schema.objectTypeToTS({ type: 'object', fields: model.fields });
      declarations.push({ type: 'interface', modelName: model.name, modelType: tsType });
    });

    return { type: 'ts', declarations };
  }
}

export function createTSGenerator(config: TSGeneratorConfig): TSGenerator {
  return new TSGeneratorImpl(config);
}
