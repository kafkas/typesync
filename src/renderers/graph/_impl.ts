import { StringBuilder } from '@proficient/ds';

import { GraphGeneration } from '../../generators/graph/index.js';
import { space } from '../../util/space.js';
import type { RenderedFile } from '../_types.js';
import type { GraphRenderer, GraphRendererConfig } from './_types.js';

class GraphRendererImpl implements GraphRenderer {
  public readonly type = 'python';

  public constructor(private readonly config: GraphRendererConfig) {}

  public async render(generation: GraphGeneration): Promise<RenderedFile> {
    const { graph } = generation;

    const b = new StringBuilder();

    b.append(`graph ${graph.orientation}` + `\n`);
    graph.links.forEach(link => {
      const [leftNodeId, rightNodeId] = link;
      b.append(space(4) + leftNodeId + ' --> ' + rightNodeId + `\n`);
    });

    return { content: b.toString() };
  }
}

export function createGraphRenderer(config: GraphRendererConfig): GraphRenderer {
  return new GraphRendererImpl(config);
}
