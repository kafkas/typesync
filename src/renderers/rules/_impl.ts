import { StringBuilder } from '@proficient/ds';
import { readFile } from 'fs/promises';

import { MisplacedStartMarkerError, MissingEndMarkerError, MissingStartMarkerError } from '../../errors/renderer.js';
import type { RulesDeclaration, RulesGeneration, RulesValidatorDeclaration } from '../../generators/rules/index.js';
import { rules } from '../../platforms/rules/index.js';
import { assertNever } from '../../util/assert.js';
import { multiply } from '../../util/multiply-str.js';
import { space } from '../../util/space.js';
import type { RenderedFile } from '../_types.js';
import type { RulesRenderer, RulesRendererConfig } from './_types.js';

class RulesRendererImpl implements RulesRenderer {
  public constructor(private readonly config: RulesRendererConfig) {}

  public async render(g: RulesGeneration): Promise<RenderedFile> {
    const b = new StringBuilder();

    const { lines, startMarkerLineIdx } = await this.preprocessOutputFile();

    lines.forEach((line, lineIdx) => {
      if (lineIdx === startMarkerLineIdx + 1) {
        const renderedDeclarations = g.declarations.map(d => this.renderDeclaration(d)).join('\n\n');
        b.append(renderedDeclarations + `\n`);
      }
      b.append(`${line}`);
      if (lineIdx !== lines.length - 1) {
        b.append('\n');
      }
    });

    const rootFile: RenderedFile = {
      content: b.toString(),
    };

    return rootFile;
  }

  private async preprocessOutputFile() {
    const { pathToOutputFile, startMarker, endMarker } = this.config;

    const outputFileContent = (await readFile(pathToOutputFile)).toString();
    const lines = outputFileContent.split('\n');
    const startMarkerLineIdx = lines.findIndex(line => line.includes(startMarker));
    const endMarkerLineIdx = lines.findIndex(line => line.includes(endMarker));

    if (startMarkerLineIdx === -1) {
      throw new MissingStartMarkerError(pathToOutputFile, startMarker);
    }

    if (endMarkerLineIdx === -1) {
      throw new MissingEndMarkerError(pathToOutputFile, startMarker);
    }

    if (startMarkerLineIdx >= endMarkerLineIdx) {
      throw new MisplacedStartMarkerError(pathToOutputFile, startMarker, endMarker);
    }

    lines.splice(startMarkerLineIdx + 1, endMarkerLineIdx - startMarkerLineIdx - 1);

    return { lines, startMarkerLineIdx };
  }

  private renderDeclaration(declaration: RulesDeclaration) {
    switch (declaration.type) {
      case 'validator':
        return this.renderValidatorDeclaration(declaration);
      default:
        assertNever(declaration.type);
    }
  }

  private renderValidatorDeclaration(declaration: RulesValidatorDeclaration) {
    const { modelName, modelType } = declaration;
    const b = new StringBuilder();
    const varName = this.config.validatorParamName;
    b.append(`${this.indent(1)}function ${this.validatorPredicate(modelName)}(${varName}) {` + `\n`);
    const predicate = rules.predicateForType(modelType, varName);
    b.append(`${this.indent(2)}return ` + this.renderPredicate(predicate) + `;\n`);
    b.append(`${this.indent(1)}}`);
    return b.toString();
  }

  private renderPredicate(predicate: rules.Predicate): string {
    switch (predicate.type) {
      case 'value-equality':
        return this.renderValueEqualityPredicate(predicate);
      case 'type-equality':
        return this.renderTypeEqualityPredicate(predicate);
      case 'type-validator':
        return this.renderTypeValidatorPredicate(predicate);
      case 'map-has-key':
        return this.renderMapHasKeyPredicate(predicate);
      case 'map-has-only-keys':
        return this.renderMapHasOnlyKeysPredicate(predicate);
      case 'literal':
        return this.renderLiteralPredicate(predicate);
      case 'or':
        return this.renderOrPredicate(predicate);
      case 'and':
        return this.renderAndPredicate(predicate);
      case 'negation':
        return this.renderNegationPredicate(predicate);
      default:
        assertNever(predicate);
    }
  }

  private renderValueEqualityPredicate(predicate: rules.ValueEqualityPredicate) {
    return `(${predicate.varName} == ${predicate.varValue})`;
  }

  private renderTypeEqualityPredicate(predicate: rules.TypeEqualityPredicate) {
    return `(${predicate.varName} is ${predicate.varType.type})`;
  }

  private renderTypeValidatorPredicate(predicate: rules.TypeValidatorPredicate) {
    return `${this.validatorPredicate(predicate.varModelName)}(${predicate.varName})`;
  }

  private renderMapHasKeyPredicate(predicate: rules.MapHasKeyPredicate) {
    return `('${predicate.key}' in ${predicate.varName})`;
  }

  private renderMapHasOnlyKeysPredicate(predicate: rules.MapHasOnlyKeysPredicate) {
    return `(${predicate.varName}.keys().hasOnly([${predicate.keys.map(k => `'${k}'`).join(', ')}]))`;
  }

  private renderLiteralPredicate(predicate: rules.LiteralPredicate) {
    return predicate.value;
  }

  private renderOrPredicate(predicate: rules.OrPredicate) {
    return `(${predicate.innerPredicates.map(p => this.renderPredicate(p)).join(' || ')})`;
  }

  private renderAndPredicate(predicate: rules.AndPredicate) {
    if (predicate.alignment === 'vertical') {
      return (
        `(\n` +
        `${predicate.innerPredicates.map(p => `${this.indent(3)}${this.renderPredicate(p)}`).join(' &&\n')}` +
        `\n${this.indent(2)})`
      );
    } else {
      return `(${predicate.innerPredicates.map(p => this.renderPredicate(p)).join(' && ')})`;
    }
  }

  private renderNegationPredicate(predicate: rules.NegationPredicate) {
    return `!${this.renderPredicate(predicate.originalPredicate)}`;
  }

  private validatorPredicate(modelName: string) {
    // TODO: Make dynamic with according to the `pattern` input
    return `isValid${modelName}`;
  }

  private indent(count: number) {
    return multiply(space(this.config.indentation), count);
  }
}

export function createRulesRenderer(config: RulesRendererConfig): RulesRenderer {
  return new RulesRendererImpl(config);
}
