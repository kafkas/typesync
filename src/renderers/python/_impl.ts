import { StringBuilder } from '@proficient/ds';

import type {
  PythonAliasDeclaration,
  PythonDeclaration,
  PythonEnumClassDeclaration,
  PythonGeneration,
  PythonPydanticClassDeclaration,
} from '../../generators/python';
import { python } from '../../platforms/python';
import { assertNever } from '../../util/assert';
import { multiply } from '../../util/multiply-str';
import { space } from '../../util/space';
import type { RenderedFile } from '../_types';
import type { PythonRenderer, PythonRendererConfig } from './_types';

class PythonRendererImpl implements PythonRenderer {
  public readonly type = 'python';

  public constructor(private readonly config: PythonRendererConfig) {}

  public async render(g: PythonGeneration): Promise<RenderedFile[]> {
    const b = new StringBuilder();

    b.append(`${this.generateImportStatements()}\n\n`);
    b.append(`${this.generateStaticDeclarations()}\n`);
    b.append(`# Model Definitions\n\n`);

    g.declarations.forEach(declaration => {
      b.append(`${this.renderDeclaration(declaration)}\n\n`);
    });

    const renderedFile: RenderedFile = {
      relativePath: this.config.rootFileName,
      content: b.toString(),
    };

    return [renderedFile];
  }

  private generateStaticDeclarations() {
    const b = new StringBuilder();
    b.append(`class TypeSyncUndefined:\n`);
    b.append(`${this.indent(1)}_instance = None\n\n`);
    b.append(`${this.indent(1)}def __init__(self):\n`);
    b.append(`${this.indent(2)}if TypeSyncUndefined._instance is not None:\n`);
    b.append(
      `${this.indent(3)}raise RuntimeError("TypeSyncUndefined instances cannot be created directly. Use UNDEFINED instead.")\n`
    );
    b.append(`${this.indent(2)}else:\n`);
    b.append(`${this.indent(3)}TypeSyncUndefined._instance = self\n\n`);
    b.append(`UNDEFINED = TypeSyncUndefined()\n`);
    return b.toString();
  }

  private generateImportStatements() {
    const b = new StringBuilder();
    b.append(`from __future__ import annotations\n\n`);
    b.append(`import typing\n`);
    b.append(`import datetime\n`);
    b.append(`import enum\n`);
    b.append(`import pydantic`);
    return b.toString();
  }

  private renderDeclaration(declaration: PythonDeclaration) {
    switch (declaration.type) {
      case 'alias':
        return this.renderAliasDeclaration(declaration);
      case 'enum-class':
        return this.renderEnumClassDeclaration(declaration);
      case 'pydantic-class': {
        return this.renderPydanticClassDeclaration(declaration);
      }
      default:
        assertNever(declaration);
    }
  }

  private renderAliasDeclaration(declaration: PythonAliasDeclaration) {
    const { modelName, modelType } = declaration;
    const expression = python.expressionForType(modelType);
    return `${modelName} = ${expression.content};`;
  }

  private renderEnumClassDeclaration(declaration: PythonEnumClassDeclaration) {
    const { modelName, modelType } = declaration;
    const b = new StringBuilder();
    b.append(`class ${modelName}(enum.Enum):\n`);
    modelType.attributes.forEach(attribute => {
      b.append(`${this.indent(1)}${attribute.key} = ${this.enumClassAttributeValueAsString(attribute)}\n`);
    });
    b.append('\n');
    return b.toString();
  }

  private enumClassAttributeValueAsString(attribute: python.EnumClassAttribute) {
    switch (typeof attribute.value) {
      case 'string':
        return `"${attribute.value}"`;
      case 'number':
        return `${attribute.value}`;
      default:
        assertNever(attribute.value);
    }
  }

  private renderPydanticClassDeclaration(declaration: PythonPydanticClassDeclaration) {
    const { modelName, modelType } = declaration;
    const b = new StringBuilder();
    b.append(`class ${modelName}(pydantic.BaseModel):\n`);
    modelType.attributes.forEach(attribute => {
      const expression = python.expressionForType(attribute.type);
      b.append(`${this.indent(1)}${attribute.name}: ${expression.content}\n`);
    });
    b.append('\n');
    return b.toString();
  }

  private indent(count: number) {
    return multiply(space(this.config.indentation), count);
  }
}

export function createPythonRenderer(config: PythonRendererConfig): PythonRenderer {
  return new PythonRendererImpl(config);
}
