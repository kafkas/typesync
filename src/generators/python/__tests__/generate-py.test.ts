import { loadSchemaForTestDefinition } from '../../../../test/util/load-schema';
import { createPythonGenerator } from '../_impl';

describe('PythonGeneratorImpl', () => {
  it('produces the correct generation for a flat schema', async () => {
    const generator = createPythonGenerator({
      indentation: 4,
      platform: 'py:firebase-admin:6',
    });
    const schema = loadSchemaForTestDefinition('flat');
    const generation = generator.generate(schema);
    expect(generation).toMatchSnapshot();
  });
});
