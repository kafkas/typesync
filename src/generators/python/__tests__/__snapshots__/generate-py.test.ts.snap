// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PythonGeneratorImpl produces the correct generation for a flat schema 1`] = `
{
  "declarations": [
    {
      "modelName": "Username",
      "modelType": {
        "name": "Placeholder",
        "type": "alias",
      },
      "type": "alias",
    },
    {
      "modelName": "UserRole",
      "modelType": {
        "name": "Placeholder",
        "type": "alias",
      },
      "type": "alias",
    },
    {
      "modelName": "Address",
      "modelType": {
        "name": "Placeholder",
        "type": "alias",
      },
      "type": "alias",
    },
    {
      "modelName": "Cat",
      "modelType": {
        "name": "Placeholder",
        "type": "alias",
      },
      "type": "alias",
    },
    {
      "modelName": "Dog",
      "modelType": {
        "name": "Placeholder",
        "type": "alias",
      },
      "type": "alias",
    },
    {
      "modelName": "Pet",
      "modelType": {
        "name": "Placeholder",
        "type": "alias",
      },
      "type": "alias",
    },
    {
      "modelName": "UserProfile",
      "modelType": {
        "name": "Placeholder",
        "type": "alias",
      },
      "type": "pydantic-class",
    },
  ],
  "type": "python",
}
`;
